### DISCUSSION TITLE
Getting IBM Z Open Editor to work with z/VM (CMS)

### DISCUSSION THREAD

**Discussion by vmdave9** [IBM/zopeneditor-about#153] [Ideas]
Wow!! This is quiet a piece of work; very impressive. It would be great to get this to work with z/VM (CMS) as well as z/OS, and we use the same languages (PL/I, Rexx, HLASM) and other tools.
I have the time to work on this if anyone there is interested.
Thanks again.
DJ
vmdave9@gmail.com

**Comment by phaumer** [comment]
@vmdave9 thanks. Could you provide some more details about what you have in mind? Which specific use cases would you like to see. With our [Zowe Explorer integration](https://ibm.github.io/zopeneditor-about/Docs/setup_integration.html) we cover a large set of z/OS use cases utilizing IBM's RSE and z/OSMF. What specifically are you missing? Any direct interactions with z/VM CMS are currently not in our scope; perhaps this is something to discuss in the Zowe project?

**Comment by vmdave9** [comment]
Hello, Peter.
Many thanks for your response back to my request. What we (the z/VM
community) are missing is the ability to seamlessly edit CMS files in the
editor....from the editor specify a file that is downloaded via FTP (or
some other method like IND$FILE), edit it with all the tools available in
the editor and then automatically save that file back to the z/VM host.  We
don't need anything like Zowe or z/OSMF as we already have other ways of
doing those functions in the z/VM environment. It's my impression and
correct me if I am wrong, that Zowe and z/OSMF are meant to be attractive,
easy-to-use, interfaces to allow folks that have little understanding of
the z/OS internal workings to be able to manage z/OS systems at least in
some sense. That's not at all what we are interested in because the z/VM
environment is orders of magnitude simpler than z/OS.
Does this help make the scope of the proposed project more acceptable? We
can discuss this in more detail if you'd like.
Take care.
Dave Jones
DJ

**Comment by vmdave9** [comment]
I forgot to add in my last note, Peter, that I am willing to do (some,
most, all?) of the work involved here, too.
DJ

**Comment by phaumer** [comment]
Hello @vmdave9 if there is an FTP server then did you try with Zowe Explorer's FTP extension? <https://marketplace.visualstudio.com/items?itemName=Zowe.zowe-explorer-ftp-extension> 
It is build on the Zowe FTP CLI plugin, which is available here: <https://github.com/zowe/zowe-cli-ftp-plugin>

**Comment by vmdave9** [comment]
Thanks, I'll check it out.

**Comment by phaumer** [comment]
Sounds good. Moving this item to the Discussions section for now.

**Comment by FALLAI-Denis** [comment]
Hi,

In addition to @phaumer's answer, using the Zowe Explorer FTP extension should allow using any remote system that has an FTP server, such as a Windows or Linux system, (Only for "USS files" functions? Certainly not for "job" functions, and see for "mvs files" functions).

It remains to be seen how Z Open Editor behaves for the remote search of COPYBOOK and INCLUDE: is it compatible with a Zowe zftp profile

**Reply by phaumer** [reply]
That is correct. FTP is not a full alternative, but over time more and more capabilities for MVS and JES were added to work in Zowe Explorer. Z Open Editor is a different story. Here we are limited to RSE API and z/OSMF, but if there are enough requests for FTP we could consider it.

**Reply by ibmrcruicks** [reply]
any update on this?  Z Open Editor ships with ZOWE VSCode extension - does the Editor make direct use of the RSE/zOSMF APIs, or does it route through the ZOWE extension?  I have been able to use the Editor seamlessly through the ZOWE FTP plugin, indicating that access is through ZOWE.  If ZOWE FTP also supported access to CMS, and provide appropriate name-mappings, it might be quite straightforward.

**Reply by phaumer** [reply]
@ibmrcruicks that is correct, we do all interactivity via the Zowe APIs that we publish in the Zowe Explorer project. Also, all the ftp capabilities are open source. The Zowe Explorer data provider is here <https://github.com/zowe/vscode-extension-for-zowe/tree/main/packages/zowe-explorer-ftp-extension> and based on the Zowe CLI FTP plugin, which itself is based on other open source libraries for FTP. So any contributions to the project(s) that increase our reach would be appreciated. If there are capabilities that you use or add to FTP that do not work with Z Open Editor specifically then we can discuss them here. 

As mentioned above, at the moment we do not have full FTP support for all features in Z Open Editor, for example, when it comes to include file resolving, user build etc. However, we left this item open as we would like to do it some time. It is a matter of priorities, which are driven by demand.

**Reply by ibmrcruicks** [reply]
thanks for the confirmation. I'll take a look at a ZOWE-compatible zvmFTP plugin that could handle sfs file/directory access, and minidisk-based sequential and maclib file types
